packages/cli/lib/ModuleManager.ts

```diff
@@ -0,0 +1,221 @@
+
+  constructor() {
+    this._applications = new Map();
+    this._plugins = new Map();
+  }
+
+  get plugins() {
+    return this._plugins;
+  }
+
+  getPlugin(type: string, name: string) {
+    if (!this._plugins.has(type)) {
+      throw new Error(pluginNotFound(name, type));
+    }
+
+    if (!this._plugins.get(type).has(name)) {
+      throw new Error(pluginNotFound(name, type));
+    }
+
+    return this._plugins.get(type).get(name);
+  }
+
+  getPluginsOfType(type: string) {
+    if (!this._plugins.has(type)) {
+      throw new Error(pluginsNotFound(type));
+    }
+
+    return this._plugins.get(type);
+  }
+
+  async prepare() {
+    for (const application of this._applications.values()) {
+      if (application.prepare) {
+        await application.prepare();
+      }
+    }
+
+    for (const plugins of this.plugins.values()) {
+      for (const plugin of plugins.values()) {
+        if (plugin.prepare) {
+          await plugin.prepare();
+        }
+      }
+    }
+  }
+
+  async cleanup() {
+    for (const application of this._applications.values()) {
+      if (application.cleanup) {
+        await application.cleanup();
+      }
+    }
+
+    for (const plugins of this.plugins.values()) {
+      for (const plugin of plugins.values()) {
+        if (plugin.cleanup) {
+          await plugin.cleanup();
+        }
+      }
+    }
+  }
+
+  async configureApplicationCommands(yargs: Yargs.Argv) {
+    for (const application of this._applications.values()) {
+      const commandModules = await application.getCommandModules();
+      for (const commandModule of commandModules) {
+        // check requirements
+        if (!commandModule.command) {
+          throw new Error("SynTest command module required command property!");
+        } else if (!commandModule.describe) {
+          throw new Error("SynTest command module required describe property!");
+        } else if (!commandModule.handler) {
+          throw new Error("SynTest command module required handler function!");
+        }
+        yargs = yargs.command(commandModule);
+      }
+    }
+
+    return yargs;
+  }
+
+  async configurePluginOptions(yargs: Yargs.Argv) {
+    for (const plugins of this.plugins.values()) {
+      for (const plugin of plugins.values()) {
+        if (plugin.getOptions) {
+          const options = await plugin.getOptions();
+
+          for (const option of options.keys()) {
+            yargs = yargs.option(`${plugin.name}-${option}`, {
+              ...options.get(option),
+              group: `${plugin.name} ${options.get(option).group || ""}`,
+            });
+          }
+        }
+      }
+    }
+    return yargs;
+  }
+
+  async getModulePath(module: string): Promise<string> {
+    let modulePath = "";
+
+    if (module.startsWith("file:")) {
+      // It is a file path
+      modulePath = path.resolve(module.replace("file:", ""));
+      if (!shell.test("-e", modulePath)) {
```

Comment:
Why do we use a shell for checking if a directory exists? Can't we use the IO libraries?

Reply:
Removed.